src/main/test/ApplicationUtilsTests.cpp

```diff
@@ -146,3 +150,197 @@ TEST_CASE("offline self-check works", "[applicationutils][selfcheck]")
+        qSet.threshold = 1;
+        qSet.validators.push_back(vNode1NodeID);
+
+        Config cfg1 = getTestConfig(1);
+        cfg1.ARTIFICIALLY_ACCELERATE_TIME_FOR_TESTING = true;
+        TmpDirHistoryConfigurator histCfg;
+        // Setup validator to publish
+        cfg1 = histCfg.configure(cfg1, /* writable */ true);
+        cfg1.MAX_SLOTS_TO_REMEMBER = 50;
+
+        Config cfg2 = getTestConfig(2);
+        cfg2.ARTIFICIALLY_ACCELERATE_TIME_FOR_TESTING = true;
+        cfg2.NODE_IS_VALIDATOR = false;
+        cfg2.FORCE_SCP = false;
+        cfg2.INVARIANT_CHECKS = {};
+        cfg2.setInMemoryMode();
+        cfg2.MODE_AUTO_STARTS_OVERLAY = false;
+        cfg2.DATABASE = SecretValue{minimalDBForInMemoryMode(cfg2)};
+        cfg2.MODE_STORES_HISTORY_LEDGERHEADERS = true;
+        // Captive core points to a read-only archive maintained by the
+        // validator
+        cfg2 = histCfg.configure(cfg2, /* writable */ false);
+
+        auto validator = simulation->addNode(vNode1SecretKey, qSet, &cfg1);
+        validator->getHistoryArchiveManager().initializeHistoryArchive(
+            histCfg.getArchiveDirName());
+
+        simulation->addNode(vNode2SecretKey, qSet, &cfg2);
+        simulation->addPendingConnection(vNode1NodeID, vNode2NodeID);
+        simulation->startAllNodes();
+
+        simulation->crankUntil(
+            [&]() { return simulation->haveAllExternalized(3, 5); },
+            10 * Herder::EXP_LEDGER_TIMESPAN_SECONDS, false);
+
+        // Generate a bit of load, and crank for some time
+        auto& loadGen = validator->getLoadGenerator();
+        loadGen.generateLoad(true, /* nAccounts */ 10, 0, 0, /*txRate*/ 1,
+                             /*batchSize*/ 1, std::chrono::seconds(0), 0);
+
+        auto& loadGenDone = validator->getMetrics().NewMeter(
+            {"loadgen", "run", "complete"}, "run");
+        auto currLoadGenCount = loadGenDone.count();
+
+        auto checkpoint =
+            validator->getHistoryManager().getCheckpointFrequency();
+
+        uint32_t selectedLedger = 0;
+        std::string selectedHash;
+
+        // Make sure validator publishes something
+        simulation->crankUntil(
+            [&]() {
+                bool loadDone = loadGenDone.count() > currLoadGenCount;
+                auto lcl =
+                    validator->getLedgerManager().getLastClosedLedgerHeader();
+                // Pick some ledger in the second checkpoint to run catchup
+                // against later
+                if (lcl.header.ledgerSeq < 2 * checkpoint)
+                {
+                    selectedLedger = lcl.header.ledgerSeq;
+                    selectedHash = binToHex(lcl.hash);
+                }
+
+                // Validator should publish at least two checkpoints
+                return loadDone &&
+                       simulation->haveAllExternalized(3 * checkpoint, 5);
+            },
+            50 * Herder::EXP_LEDGER_TIMESPAN_SECONDS, false);
+
```

Comment:
Maybe require selectedLedger and selectedHash got populated here?

Reply:
Good idea, done.